src/server/scene/basic_idle_hub.cpp

```diff
@@ -0,0 +1,246 @@
+    Multiplexer()
+        : ObserverMultiplexer{system_executor}
+    {
+    }
+
+    void idle() override
+    {
+        for_each_observer(&IdleStateObserver::idle);
+    }
+
+    void active() override
+    {
+        for_each_observer(&IdleStateObserver::active);
+    }
+};
+
+ms::BasicIdleHub::BasicIdleHub(
+    std::shared_ptr<time::Clock> const& clock,
+    mt::AlarmFactory& alarm_factory)
+    : clock{clock},
+      alarm{alarm_factory.create_alarm(
+          std::make_unique<AlarmCallback>(mutex, [this](std::unique_lock<std::mutex>& lock)
+            {
+                alarm_fired(lock);
+            }))}
+{
+    poke();
+}
+
+ms::BasicIdleHub::~BasicIdleHub()
+{
+    alarm->cancel();
+}
+
+void ms::BasicIdleHub::poke()
+{
+    std::unique_lock<std::mutex> lock{mutex};
+    poke_time = clock->now();
+    schedule_alarm(lock, poke_time);
+    if (!idle_multiplexers.empty())
+    {
+        auto const idle = std::move(idle_multiplexers);
+        idle_multiplexers.clear();
+        lock.unlock();
+        for (auto const& multiplexer : idle)
+        {
+            multiplexer->active();
+        }
+    }
+}
+
+void ms::BasicIdleHub::register_interest(
+    std::weak_ptr<IdleStateObserver> const& observer,
+    time::Duration timeout)
+{
+    register_interest(observer, system_executor, timeout);
+}
+
+void ms::BasicIdleHub::register_interest(
+    std::weak_ptr<IdleStateObserver> const& observer,
+    Executor& executor,
+    time::Duration timeout)
+{
+    std::unique_lock<std::mutex> lock{mutex};
+    auto const iter = timeouts.find(timeout);
+    bool is_active = true;
+    if (iter == timeouts.end())
+    {
+        auto const multiplexer = std::make_shared<Multiplexer>();
+        multiplexer->register_interest(observer, executor);
+        timeouts.insert({timeout, multiplexer});
+        first_timeout = timeouts.begin()->first;
+        auto const current_time = clock->now() - poke_time;
+        if (alarm_timeout && alarm_timeout.value() <= timeout)
+        {
+            // The alarm will be fired before we hit our timeout
+            is_active = true;
+        }
+        else if (current_time < timeout)
+        {
+            // The alarm will be fired after we hit our timout,
+            // but we have not hit it yet so the alarm needs to be moved up
+            is_active = true;
+            alarm_timeout = timeout;
+            alarm->reschedule_in(std::chrono::duration_cast<std::chrono::milliseconds>(timeout - current_time));
+        }
+        else
+        {
+            // This timeout has already been passed
+            is_active = false;
+            idle_multiplexers.push_back(multiplexer);
+        }
+    }
+    else
+    {
+        iter->second->register_interest(observer, executor);
+        is_active = (alarm_timeout && alarm_timeout.value() <= timeout);
+    }
+    lock.unlock();
+    if (auto const shared = observer.lock())
+    {
+        if (is_active)
+        {
+            shared->active();
```

Comment:
Oh! This isn't dispatched on the `executor`, like other all observations will be. It probably should be?